domain/resource/service/resource.go

```diff
@@ -163,6 +175,28 @@ func (s *Service) GetApplicationResourceID(
 	return s.st.GetApplicationResourceID(ctx, args)
 }
 
+// GetResourceUUIDByApplicationAndResourceName returns the ID of the application
+// resource specified by natural key of application and resource name.
+//
+// The following error types can be expected to be returned:
+//   - [resourceerrors.ApplicationNotFound] is returned if the application is
+//     not found.
+//   - [resourceerrors.ResourceNotFound] if no resource with name exists for
+//     given application.
+func (s *Service) GetResourceUUIDByApplicationAndResourceName(
+	ctx context.Context,
+	appName string,
+	resName string,
+) (coreresource.UUID, error) {
+	if strings.TrimSpace(resName) == "" {
+		return "", resourceerrors.ResourceNameNotValid
+	}
+	if strings.TrimSpace(appName) == "" {
+		return "", resourceerrors.ApplicationNameNotValid
+	}
```

Comment:
issue:  strings.TrimSpace is not sufficient validation of application nor resource names.

It's already been done in the application domain here:
https://github.com/juju/juju/blob/d986ad9bb76cfedc5191ea5a0edba9bbdd2590f1/domain/application/service/service.go#L352-L355

You can do something similar for resources.

Correct regex for these can be found in the names package.

Reply:
Nice, more validation is better. I've added one for relations, but I can't see one anywhere in the names package or 3.6 for resources. The only validation that takes place is checking that it is not empty: https://github.com/juju/charm/blob/v12/resource/meta.go#L39